llama-cpp-2/src/context/session.rs

```diff
@@ -130,4 +130,29 @@ impl LlamaContext<'_> {
             }
         }
     }
+
+    /// Returns the maximum size in bytes of the state (rng, logits, embedding
+    /// and kv_cache) - will often be smaller after compacting tokens
+    pub fn get_state_size(&self) -> usize {
+        unsafe { llama_cpp_sys_2::llama_get_state_size(self.context.as_ptr()) }
+    }
+
+    /// Copies the state to the specified destination address.
+    /// Destination needs to have allocated enough memory.
+    /// Returns the number of bytes copied
+    pub unsafe fn copy_state_data(&self, dest: *mut u8) -> usize {
+        unsafe {
+            llama_cpp_sys_2::llama_copy_state_data(self.context.as_ptr(), dest)
+        }
+    }
+
+    /// Set the state reading from the specified address
+    /// Returns the number of bytes read
+    pub unsafe fn set_state_data(&mut self, src: &[u8]) -> usize {
+        unsafe {
+            // we don't really need a mutable pointer for `src` -- this is a llama-cpp lapse,
+            // so we cast away the constness
+            llama_cpp_sys_2::llama_set_state_data(self.context.as_ptr(), src.as_ptr() as *mut u8)
+        }
+    }
```

Comment:
make `src` mutable. I'll open a PR to llama.cpp to make it `const` on their end

Reply:
you can document that `src` will be changed to be immutable pending upstream changes, I'll fix once PR is merged.

Reply:
or wait until PR is merged and the nightly update llama.cpp goes though.